### PR TITLE
separete lane connections for track/road vehicles.

### LINKED ISSUE #354: Lane connector: Separate road vs. tracked connections
### Description

Using lane connector tool, tram track lanes can only be connected to other tram track lanes.

However, in most cases the tram track lane is also a road or bus lane. It would be nice if we could connect the non-tram traffic to any other traffic lane.

Maybe we could have two shapes on incoming lanes:

* Circle = the connector for road vehicles
* Rectangle = connector for tracked vehicles

This way we could still allow road vehicles to connect to any outgoing road lane, and the tracked (eg. tram, usually) vehicle would be limited to outgoing lanes containing same type of track.

### Screenshots (Optional)

Example of a road that has tram+bus lanes - it's not possible to connect those bus lanes to the bus lanes on adjoining road.

255710_screenshots_20190601215741_1

tasks:
- [x] separate routing for car/track vehicles in routing manager #1501 #1493 #1477 
- [x] sub lane connection manager #1545 #1546 
- [ ] use different sub managers for road/track vehicles
- [ ] (optional) handle trolleybuses

### LINKED ISSUE #1198: TMPE connections is not restablished after removing lane connection
> Milestone update: This was partially fixed in v11.6.2 as shown in changelog, but [additional issue was identified](https://github.com/CitiesSkylinesMods/TMPE/issues/1198#issuecomment-1127964194) which will be fixed by #1550 in v11.6.6.0. As such the milestone for this issue has been updated to 11.6.6.0.

I tested TMPE lane connections on Train Network using AR(Adaptive roads) mod and HUT mod(Hide unconnected tracks).
AR gets connections from Routing Manger
HUT gets connections from Lane Connection Manager.

when I create remove the red connection
image
AR shows this : tracks not reconnected which is indicates a bug:
image
HUT shows this : tracks VISUALLY (But not really) reconnected as expected:
image

This means Lane Connection manager rightfully removes the lane connection data but routing manager does not.

To prove that the problem is with TMPE and not with AR I tested it with train stops:
image
in the picture above I cannot make a stop which means even though I removed the lane connection, The tracks did not get reconnected.

Note: If I force node update using MoveIT or if I create another lane connection on another segment, the track data will update correctly.

Note2: when I add lane connection on train tracks, I notice in the log that LaneArrowManger is updated. but when I remove tracks LaneArrowManager is not updated. I am not sure if LaneArrowManager is relevant for train networks or if this is somehow relevant.

Need: routing manager should reconnect tracks after I remove lane connection

### PR DESCRIPTION
**Pull request by kianzarrin**
fixes #354 

* hotkeys and icons from #1479
* trolleybuses use same connections as cars.
* used Road/Track sub managers to store lane connections.
* all other data structures need to use separate arrays to store road/track connections.
* lane connection manager does not allow making unsupported lane connection of the wrong lane/vehicle type. this helps with backward compatibility too. but it does allow making backward connections when not allowed for the sake of backward compatibility on train stations. this should change in #1213.

Bonus:
* fixes #1198 : update connections when new connection is made.
* update direction arrow texture (less aliasing).

[TMPE.zip](https://ci.appveyor.com/api/projects/krzychu124/tmpe/artifacts/TMPE.zip?branch=354-phase3)

pictures are in the comments.

Tests:
- [x] car only test: always in car mode. holding alt/shift does not change mode
- [x] track only test: always in track mode. holding alt.shift does not change mode
- [x] other node test: regardless of whether or not current node is car-only or track-only none-selected nodes show all car+track connections. if alt is held then they only show track connections.
- [x] tram road test: holding alt make tram roads select-able. otherwise only car lanes are select-able.
- [x] mixed lane mode: normally you see circle and can make car connection. holding alt shows triangle and can make track connection. holding shift can make both connections.
- [x] bidirectional lane: node marker is diamond and you can make forward/backward connection. lane curves have arrows on them indicating their direction.
- [x] acute+mixed test: hold shift to select mixed lanes. acute targets should show circle, and clicking them should only make car lane connection
- [x] acute only mixed lane test: connect a road with mixed tram lane to an intersection with no other tram rods. expected: no 
target lane connection is highlighted. shift clicking will put in car mode.
- [x] source to target test: make sure you can only make lane connection in the right direction.
- [x] hintbox text: go through different state and check if hintbox is OK. make sure ALT/SHIFT modifiers are only there when selecting car+tram node.
- [x] big roundabout test: make sure roundabout connector makes stay in lane is applied on entry and inside of the roundabout when there is a node too close to the roundabout and when there are nodes inside roundabout. related #705
- [x] mixed lane arrow test: connect trams only on a mixed lane and make sure you can still modify lane arrows. make sure that you can't modify lane arrows if you connect cars only. 
- [x] separate turning lane: in above scenario make sure separate turning lanes is blocked iff there is a car connection.

mixed connections tests: connect car and track differently on a mixed car+tram lane and:
- [x] mixed lane routing test: verify the are routed differently. 
- [x] refresh test: exit and reenter tool, select the node and make sure you can see all connections.
- [x] lane arrow test: lane arrows are influenced only by car connection.
- [x] mixed lane persistency test: make sure they connections persist when save/load game
- [x] mixed lane record test: make sure connections can be copied by move it.

backward compatibility test:
- [x] save game backward compatibility test: connect mixed lanes in old TMPE and saved it. make sure they get both car/track connection when loaded with this version. 
- [x] lane record backward compatibility test: connect mixed lanes in old TMPE and export it. make sure they get both car/track connection when imported with this version. 

TODO:
- [X] on mixed lanes do not allow trams to make acute lane connections.
- [x] hints

### CODE REVIEW

**Comment by kianzarrin** [comment]
image
image
image

**Comment by aubergine10** [comment]
Suggestion: Concise tooltips when mouse hovers over one of the shapes to help user understand what the shape represents.

**Comment by kvakvs** [comment]
Looks nice
Make sure that the textures you draw have enough resolution to not be more blurry than the lane sausages and the small circles, so that it looks nice together. Currently on screenshots the triangles are more blurry and standing out.

**Comment by aubergine10** [comment]
> Currently on screenshots the triangles are more blurry and standing out.

To me it looks like the triangles are fine, but the black border around them is thinner than the black border around the circles, so it's maybe just the black border that's looking blurry?

**Comment by kianzarrin** [comment]
> Currently on screenshots the triangles are more blurry and standing out.

Both triangle and the diamond use the same texture: 256x256 white (the quad gives it shape). So I don't think its a resolution issue. I don't notice any blur. maybe that is just how the overlay shader works?

> but the black border around them is thinner than the black border around the circles

You are right. I expand them.

**Comment by kianzarrin** [comment]
@kvakvs @krzychu124 @aubergine10 
I was wondering maybe I can draw squares for uni-directional tracks and diamonds for bi-directional tracks.
triangles are not easy to work with:
  - they are small (half the size of a square)
  - they don't cover the lane end and lane curve very well (unless if I shift them but that make the code looks hackish.
  - their shadow needs re-centering (triangles are asymmetrical.

**Comment by kianzarrin** [comment]
Enlarged triangle outline 
image

**Comment by kvakvs** [comment]
I think you do not need special shape for single/bidirectional lane ends, one shape is fine. Hell even circle is fine. A shape shows the traffic source, so shape on both ends of a junction means traffic comes from both directions.

Since you're using a PNG file it should make no difference in your code, whatever shape you draw in photoshop, even a 🍆 will work just fine, so triangles will work too. But the black outline is still very blurry.

**Comment by kianzarrin** [comment]
> @kvakvs Since you're using a PNG file it should make no difference in your code, whatever shape you draw in photoshop, even a 🍆 will work just fine, so triangles will work too. But the black outline is still very blurry.

1- Centering stuff does matter.
2- I am using square. because triangle will look ugly:
https://media.discordapp.net/attachments/968969871884951572/969000601235111936/unknown.png

**Comment by kianzarrin** [comment]
I ditched triangle and am using square now. Triangle looks a bit ugly and is hard to write the code.
image

**Comment by aubergine10** [comment]
Looks good!

**Comment by kianzarrin** [comment]
about backward compatibility at train stations:
In TMPE stable one lane connection like bellow is enough to make trains stay in lane.
image
The same is true with TMPE in this PR:
image

This makes backward compatibility easy. There is no need to keep connections that from target to source when loading legacy game.

Removing connections in the wrong direction also solves this issue that happens when you remove lane connections that was made using legacy TMPE:
image

**Comment by krzychu124** [comment]
Sync with `master` :)

File: TLM/TLM/State/Configuration.cs
Comment (Elesbaan70) [inline TLM/TLM/State/Configuration.cs:190]: I am concerned about continuing to introduce breaking file format changes. Can this all wait until the new persistence code is merged?

Comment (kianzarrin) [reply TLM/TLM/State/Configuration.cs:190]: This PR has waited for a month already. I don't think we need to wait more. conflicts will accumulate over time.
Is it breaking anything?

Comment (kianzarrin) [reply TLM/TLM/State/Configuration.cs:190]: it does not break anything for me. I made the connections bellow with TMPE from this PR then saved and loaded the game using stable TMPE:
image

Comment (kianzarrin) [reply TLM/TLM/State/Configuration.cs:190]: If we don't want to merge it with master, we should consider merging a feature branch instead of stopping this PR. but it is forward/backward compatible so I don't think we should.

Comment (krzychu124) [reply TLM/TLM/State/Configuration.cs:190]: It's exactly the same situation as last time. If you save game with this build you won't be able to load it using current `master` or WS stable

Comment (krzychu124) [reply TLM/TLM/State/Configuration.cs:190]: ok, did not throw exception or anything, loaded something, looks good. Ignore previous comment

Comment (kianzarrin) [reply TLM/TLM/State/Configuration.cs:190]: > loaded something

it is both forward/backward compatible. 
if you make different lane connections on mixed/car track lanes -> save -> load with earlier version of TMPE, then car/track lane connections will merge.

Comment (krzychu124) [reply TLM/TLM/State/Configuration.cs:190]: yeah, noticed that, looks ok.

**Review by krzychu124 — APPROVED** [review]

**Review by chameleon-tbn — APPROVED** [review]
lgtm

**Pull request merged**
